pyccel/codegen/printing/cucode.py

```diff
@@ -61,14 +61,45 @@ def _print_Module(self, expr):
             else:
                 local_imports += self._print(imp)
 
-        imports = f'{c_headers_imports}\
-                    extern "C"{{\n\
-                    {local_imports}\
-                    }}'
+        imports = f'{c_headers_imports}\n\
+                    {local_imports}\n'
 
         code = f'{imports}\n\
                  {global_variables}\n\
                  {body}\n'
 
         self.exit_scope()
         return code
+
+    def _print_ModuleHeader(self, expr):
+        self.set_scope(expr.module.scope)
+        self._in_header = True
+        name = expr.module.name
+        if isinstance(name, AsName):
+            name = name.name
+        # TODO: Add interfaces
+
+        funcs = ""
+        cuda_headers = ''
+        for f in expr.module.funcs:
+            if not f.is_inline:
+                if 'kernel' in f.decorators:  # Checking for 'kernel' decorator
+                    cuda_headers += self.function_signature(f) + ';\n'
+                else:
+                    funcs += self.function_signature(f) + ';\n'
+        global_variables = ''.join('extern '+self._print(d) for d in expr.module.declarations if not d.variable.is_private)
+        # Print imports last to be sure that all additional_imports have been collected
+        imports = [*expr.module.imports, *self._additional_imports.values()]
+        imports = ''.join(self._print(i) for i in imports)
+
+        self._in_header = False
+        self.exit_scope()
+        imports = f'{cuda_headers}\n\
+                    extern "C"{{\n\
+                    {funcs}\n\
+                    }}'
+        return (f"#ifndef {name.upper()}_H\n \
+                #define {name.upper()}_H\n\n \
+                {global_variables}\n \
+                {imports}\n \
+                #endif // {name}_H\n")
```

Comment:
Shouldn't this also be `name.upper()`?
```suggestion
                #endif // {name.upper()}_H\n")
```